lib/charms/sdcore_nrf/v0/fiveg_nrf.py

```diff
@@ -0,0 +1,296 @@
+
+    def _on_nrf_available(self, event: NRFAvailableEvent):
+        nrf_url = event.url
+        <do something with the nrf_url>
+
+
+if __name__ == "__main__":
+    main(DummyFiveGNRFRequirerCharm)
+```
+
+### Provider charm
+The provider charm is the one requiring providing the information about the NRF.
+
+Example:
+```python
+
+from ops.charm import CharmBase, RelationJoinedEvent
+from ops.main import main
+
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFProvides
+
+
+class DummyFiveGNRFProviderCharm(CharmBase):
+
+    NRF_URL = "https://nrf.example.com"
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.nrf_provider = NRFProvides(self, "fiveg-nrf")
+        self.framework.observe(
+            self.on.fiveg_nrf_relation_joined, self._on_fiveg_nrf_relation_joined
+        )
+
+    def _on_fiveg_nrf_relation_joined(self, event: RelationJoinedEvent):
+        if self.unit.is_leader():
+            self.nrf_provider.set_nrf_information(
+                url=self.NRF_URL,
+                event=event
+            )
+
+    def _on_nrf_url_changed(
+        self,
+    ):
+        self.nrf_provider.set_nrf_information(
+            url="https://different.nrf.com",
+            update_all_relations=True,
+        )
+
+
+if __name__ == "__main__":
+    main(DummyFiveGNRFProviderCharm)
+```
+
+"""
+
+import logging
+from typing import Optional, Union
+
+from interface_tester.schema_base import DataBagSchema  # type: ignore[import]
+from ops.charm import CharmBase, CharmEvents, RelationChangedEvent, RelationJoinedEvent
+from ops.framework import EventBase, EventSource, Handle, Object
+from ops.model import Relation
+from pydantic import AnyHttpUrl, BaseModel, Field, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "cd132a12c2b34243bfd2bae8d08c32d6"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = [
+    "pydantic" "pytest-interface-tester",
+]
+
+
+logger = logging.getLogger(__name__)
+
+"""Schemas definition for the provider and requirer sides of the `fiveg_nrf` interface.
+It exposes two interfaces.schema_base.DataBagSchema subclasses called:
+- ProviderSchema
+- RequirerSchema
+Examples:
+    ProviderSchema:
+        unit: <empty>
+        app: {"url": "https://nrf-example.com:1234"}
+    RequirerSchema:
+        unit: <empty>
+        app:  <empty>
+"""
+
+
+class ProviderAppData(BaseModel):
+    """Provider app data for fiveg_nrf."""
+
+    url: AnyHttpUrl = Field(
+        description="Url to reach the NRF.", examples=["https://nrf-example.com:1234"]
+    )
+
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for fiveg_nrf."""
+
+    app: ProviderAppData
+
+
+def data_is_valid(data: Union[str, dict]) -> bool:
```

Comment:
This method name is ambiguous. `data` is very generic. Something like `data_matches_provider_schema` would be more precise.

Reply:
Also, why accept either a string or a dict. It makes it more bug prone.

Reply:
Method name changed to `data_matches_provider_schema`. 
About accepting both `str` and `dict`, that was the tradeoff of making it generic for both provider and requirer classes, following the advice in this [comment](https://github.com/canonical/sdcore-nrf-operator/pull/6#discussion_r1191748635). Let me know if that makes sense.

Reply:
I'm fine with the method being generic, it doesn't mean though that it must accept both a string and a dict. One of the 2 can do some manipulation before calling `data_matches_provider_schema`. 

Example:
```python
from interface_tester.schema_base import DataBagSchema
from pydantic import AnyHttpUrl, BaseModel, Field, ValidationError


class ProviderAppData(BaseModel):
    url: AnyHttpUrl = Field(
        description="Url to reach the NRF.", examples=["https://nrf-example.com:1234"]
    )


class ProviderSchema(DataBagSchema):
    app: ProviderAppData


def data_matches_provider_schema(data: dict) -> bool:
    try:
        ProviderSchema(app=data)
        return True
    except ValidationError:
        return False


url="http://1.2.3.4"
data_matches_provider_schema(data={"url": url})
```

Reply:
It felt weird doing manipulation at first, but definitely makes sense. Changed!